python/revng/api/_capi.py

```diff
@@ -26,7 +51,10 @@ def __init__(self, api, ffi: FFI):
         self.__api = api
         self.__ffi = ffi
         self.__lock = Lock()
+        self.__counter = AtomicCounterWithCallback(self.__api.rp_shutdown)
         self.__proxy: Dict[str, Callable[..., Any]] = {}
+        # We're using self.__dict__ to bypass __setattr__
+        self.__dict__["close"] = lambda: self.__counter.mark_end()
```

Comment:
Uh?

Reply:
It was a bit of python magic to bypass `__setattr__`, however I've had a second look and reworked it without the magic